000-mm2-connector-offsets-support.md

```diff
@@ -0,0 +1,449 @@
+* Monitoring the progress of running connectors on a per-source-partition basis 
+* Skipping records that cause issues with a connector and cannot be addressed using existing error-handling features
+
+## Proposal
+
+Strimzi should include mechanisms to let users manage offsets when utilising the KafkaMirrorMaker2 resources.
+Many users of MirrorMaker make use of this resource, rather than the KafkaConnector resource, so it's important to support it here as well.
+Similar, to [proposal #76][proposal-76], this proposal suggests introducing a new annotation on the KafkaMirrorMaker2 resource to trigger each operation.
+The operations are:
+* List offsets.
+* Alter offsets.
+* Reset offsets.
+
+The list offsets operation uses a ConfigMap for the output.
+The alter offsets operation uses a ConfigMap for the input.
+
+In addition, a separate new annotation is proposed to allow the user to select which MirrorMaker connector to apply the action to.
+
+### Request/Response from the API endpoints
+
+The response body received when using the GET endpoint, and the request body used when calling the PATCH endpoint have identical structures to allow users to easily fetch and then alter offsets.
+The structure of the body is not fully defined for generic source connectors.
+This is because source connectors determine their own structure for offsets depending on the source system they are integrating with.
+However, for each of the MirrorMaker connectors (MirrorSourceConnector, MirrorCheckpointConnector, MirrorHeartbeatConnector) we know the expected structure.
+
+MirrorSourceConnector format:
+```
+{
+  "offsets": [
+    {
+      "partition": {
+        "cluster": "east-kafka",
+        "partition": 0,
+        "topic": "mirrormaker2-cluster-configs"
+      },
+      "offset": {
+        "offset": 0
+      }
+    }
+  ]
+}
+```
+
+MirrorCheckpointConnector format:
+```
+{
+  "offsets": [
+    {
+      "partition": {
+        "partition": 4,
+        "topic": "inventory",
+        "group": "my-group"
+      },
+      "offset": {
+        "offset": 0
+      }
+    }
+  ]
+}
+```
+
+MirrorHeartbeatConnector format:
+```
+{
+  "offsets": [
+    {
+      "partition": {
+        "sourceClusterAlias": "east-kafka",
+        "targetClusterAlias": "west-kafka"
+      },
+      "offset": {
+        "offset": 0
+      }
+    }
+  ]
+}
+```
+
+### New annotations
+
+All three actions will make use of a new annotation called `strimzi.io/connector-offsets`.
+The possible values will be `list`, `alter`, and `reset`.
+
+A second new annotation will be added called `strimzi.io/mirrormaker-connector`.
+It will be required when `strimzi.io/connector-offsets` is set.
+The value will be the name of the connector to apply the action to, for example `east-kafka->west-kafka.MirrorSourceConnector`.
```

Comment:
Because the JSON field in the output won't have the -> anyway but it's replaced by --, does it really make sense to have the name this way? Maybe having the -- or just - since the beginning as the annotation value? I am just worried that for users and support we have to refer to this replacement and clarify it multiple times.

Reply:
I think it might be useful to note the naming convention you mentioned in a later comment: _Strimzi doesn't allow users to specify the connector names for MirrorMaker. It puts the name as SOURCE_ALIAS->TARGET_ALIAS.CONNECTOR_

Reply:
@ppatierno so you're suggesting that the user provides the `--` version of the connector name for both the annotation and the entry in the ConfigMap? I think that does make sense, then they don't have to remember which is which. @scholzj does that change make sense to you?

Reply:
TBH, I don't that makes sense and I think that would be way more confusing. I would keep the correct name as MM2 uses it.

Reply:
@ppatierno Any chance you could get back to this ^^^? You are the only one who seems to have some open comments. It would be great to have things resolved to be able to merge it.

Reply:
I think what I missed here is that the format with the `->` is the "official" one used by KafkaConnect, right?
If now my understanding is correct, then I think the current proposal makes more sense.